### PR TITLE
[torch.export] Pattern Matching Using Literal Arguments

### PR DESCRIPTION
**Pull request by rvijayc**
# Problem

We have a PT2 based workflow where we use [replace_pattern](https://github.com/pytorch/pytorch/blob/a7509e98c5519c30613991f8593ae43454c6f8be/torch/fx/subgraph_rewriter.py#L105) to match patterns of ATen Ops and replace it with our custom Ops. The patterns and replacements are compiled using `torch.export`.  We also use the same patterns with [SubgraphMatcherWithNameNodeMap](https://github.com/pytorch/pytorch/blob/fb934629040ae98d9538dfaed99ad1ce66f3bdc2/torch/fx/passes/utils/matcher_with_name_node_map_utils.py#L44) for quantization annotation in the [PT2 Export Quantization Workflow](https://pytorch.org/tutorials/prototype/pt2e_quant_ptq.html).

During this process, we encounter issues where we are not able to pattern match literal arguments as they get "burned-in" in the compiled pattern (which is actually the expected behavior of `torch.export`). 

Here's one such example of a pattern ...

```python
class LayerNormPattern(torch.nn.Module):

    def forward(self, x: torch.Tensor, normalized_shape: List[int], 
                weight: torch.Tensor, bias: torch.Tensor, 
                eps: float) -> torch.Tensor:

        # match LayerNorm
        y = torch.nn.functional.layer_norm(x, normalized_shape, weight, bias, eps)

        # ... optionally, match other functionality ...

        return y
```

... and the corresponding replacement (just an example stub to illustrate the concept).

```python
@torch.library.custom_op("my_lib::my_layer_norm", mutates_args=())
def my_layer_norm(x: torch.Tensor, normalized_shape: List[int],
                  weight: torch.Tensor, bias: torch.Tensor, 
                  eps: float) -> torch.Tensor:

    # for now, returns corresponding ATEN operator output, but could contain a
    # customized version of layer_norm.
    y = torch.ops.aten.layer_norm(x, normalized_shape, weight, bias, eps)

    # ... other optional functionality ...

    return y

@torch.library.register_fake("my_lib::my_layer_norm")
def _(x: torch.Tensor, normalized_shape: List[int], 
      weight: torch.Tensor, bias:torch.Tensor, 
      eps: float):

    return torch.empty_like(x)

class LayerNormReplacement(torch.nn.Module):

    def forward(self, x: torch.Tensor, normalized_shape: List[int], 
                weight: torch.Tensor, bias: torch.Tensor, 
                eps: float) -> torch.Tensor:

        # match LayerNorm
        y = torch.ops.my_lib.my_layer_norm(x, normalized_shape, weight, bias, eps)

        # ... optionally, match other functionality ...

        return y
```

When the pattern is compiled using `torch.export`, it produces a graph like below:

image

The literals (the `normalized_shape` array and `eps`) do show up in the graph, but are disconnected and represent "dead code". Additionally, the `normalized_shape` array argument is replicated across each array index which isn't desired either. Ideally, I would like the compiled pattern to look like this:

image

The original patterns generated by `torch.export` are, as such, unusable with [SubgraphMatcher]() or [replace_pattern](https://github.com/pytorch/pytorch/blob/a7509e98c5519c30613991f8593ae43454c6f8be/torch/fx/subgraph_rewriter.py#L105). We'll simply error out with <em>"SubgraphMatcher cannot be initialized with an pattern with dead code"</em> before we get any further.

# Solution

There are a few ways to overcome this today:

1. Define tensor overload for literals. This is typically used for literals that represent scales and zero points. The torch quantize/dequantize ops have overload for these quantities. This doesn't scale very well as it needs to be done for every literal in every op.
2. Use a helper function like [_replace_literals_with_existing_placeholders](https://github.com/pytorch/pytorch/blob/fb934629040ae98d9538dfaed99ad1ce66f3bdc2/torch/ao/quantization/pt2e/utils.py#L528), but it has a restriction where all the literals must have unique values. It is not easy to meet this condition in practice. For example, if a op has three boolean arguments, we are out of luck.

What I am proposing this a third option for explicit conversion of literals to placeholder by simply providing a mapping specification. Below is an example of how it will be used:

First, define a mapping between the Op, its arguments, and the corresponding placeholder names.

```python
#
# Define the mapping between a node's arguments and its corresponding placeholder nodes.
#
LN_REPL_LITERAL_MAP = {
    'my_lib::my_layer_norm': {     # the node (call_function) for which this replacement applies
        1: 'normalized_shape',     # <node_arg_idx> -> <op_placeholder_name> mapping.
        4: 'eps',
    }
}
```

Call `_replace_node_literals_with_existing_placeholders` with the EXIR graph and the literal mapping to replace literals with placeholders as follows:

```python
pattern_gm = torch.export.export(pattern_model, cls.example_inputs).module()
_replace_node_literals_with_existing_placeholders(pattern_gm, LN_PATTERN_LITERAL_MAP)
```

This will correct the EXIR graph and attach the placeholder nodes using the mapping define `LN_PATTERN_LITERAL_MAP`. Here's the output graph:

```
opcode         name              target                   args                                      kwargs
-------------  ----------------  -----------------------  ----------------------------------------  --------
placeholder    x                 x                        ()                                        {}
placeholder    normalized_shape  normalized_shape         ()                                        {}
placeholder    weight            weight                   ()                                        {}
placeholder    bias              bias                     ()                                        {}
placeholder    eps               eps                      ()                                        {}
call_function  layer_norm        aten.layer_norm.default  (x, normalized_shape, weight, bias, eps)  {}
output         output_1          output                   ((layer_norm,),)                          {}
```

Ideally, it would be nice if `torch.export` took an option to treat literal arguments to a module as placeholders, but barring that, this is a workaround.

If there are easier options to achieve this, or if I missed anything, please let me know.

### CODE REVIEW

**Review by eellison — COMMENTED** [review]
@SherlockNoMad , we should have a discussion about the dual maintenance of this and inductor patcher. is it worth maintaining both ?
